src/Functions/GatherUtils/Algorithms.h

```diff
@@ -671,13 +678,42 @@ void NO_INLINE arrayAllAny(FirstSource && first, SecondSource && second, UInt8 *
     }
 }
 
+template <typename Source, typename Sink>
+inline void checkAreLowCardinalityInsertable(const Source &, const Sink &)
+{
+}
+
+template <typename Source>
+requires std::same_as<typename Source::Slice, GenericArraySlice>
+inline void checkAreLowCardinalityInsertable(const Source & source, const GenericArraySink & sink)
+{
+    const IColumn & source_elements = source.elements;
+
+    const bool can_insert =
+        // Check same type
+        source_elements.structureEquals(sink.elements)
+            // insertRangeFrom casts numeric types if they are different, so it can insert them.
+            || (sink.elements.lowCardinality()
+                && source_elements.lowCardinality()
+                && sink.elements.isNumeric()
+                && source_elements.isNumeric());
+
+    if (!can_insert)
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Function resizeDynamicSize expects same column types or numeric lowCardinality for GenericArraySlice and GenericArraySink.");
+}
+
+
 template <typename ArraySource, typename ValueSource, typename Sink>
 void resizeDynamicSize(ArraySource && array_source, ValueSource && value_source, Sink && sink, const IColumn & size_column)
 {
     const auto * size_nullable = typeid_cast<const ColumnNullable *>(&size_column);
     const NullMap * size_null_map = size_nullable ? &size_nullable->getNullMapData() : nullptr;
     const IColumn * size_nested_column = size_nullable ? &size_nullable->getNestedColumn() : &size_column;
 
+    checkAreLowCardinalityInsertable(array_source, sink);
```

Comment:
The "generic" implementation should already check that the types are identical, and the calling code should convert to the identical types before going here.